lib/api/api.js

```diff
@@ -184,14 +186,115 @@ const api = {
             }
             return { returnTagCount, isImplicitDeny };
         }
+        let formDataParser;
+        let fileEventData = null;
+
+        if (apiMethod === 'objectPost') {
+            if (request.headers['content-type'].includes('multipart/form-data')) {
+                formDataParser = busboy({ headers: request.headers });
+            } else {
+                const contentTypeError = errors.PreconditionFailed
+                    .customizeDescription('Bucket POST must be of the enclosure-type multipart/form-data');
+                return process.nextTick(callback, contentTypeError);
+            }
+        }
 
         return async.waterfall([
+            next => {
+                if (apiMethod === 'objectPut' || apiMethod === 'objectPutPart') {
+                    return next(null);
+                }
+                if (apiMethod === 'objectPost' && request.headers['content-type'].includes('multipart/form-data')) {
```

Comment:
When we receive a post we should maybe also have a timeout or an interval to terminate the request and free memory if the request takes too much time to be written. 

We could have a test with a slow request.

Reply:
Here is a note about request timeout from the readme we might want to take into account for node 18  (https://github.com/mscdex/busboy?tab=readme-ov-file)

> Note: If you are using node v18.0.0 or newer, please be aware of the node.js HTTP(S) server's [requestTimeout](https://nodejs.org/docs/latest/api/http.html#httpcreateserveroptions-requestlistener) configuration setting that is now enabled by default, which could cause upload interruptions if the upload takes too long.

Reply:
I wonder if this is handled in the fastify busboy package in some way... I can't imagine the parser will have any control over that actually. Maybe we have to disable the timeout somewhere in the code ?

Reply:
Yep busboy is for parsing. We have to manage timeout on our own.
Maybe test on AWS with a slow request to see how much you can keep it open.

Reply:
In my experience as long as the connection is maintained, AWS does not timeout, having tested with 5Gb files that took quite a while to upload

Reply:
> But for the POST it seems the timeout goes to 40s for fields before the file field

With the limit to 20kb for field data, how were you able to test this?

Reply:
With netcat:

If I put a bigger Content-Length, and leave the body unfinished:

```bash
cat <<"EOF" | time nc -C micktestpost.s3.eu-north-1.amazonaws.com 80
POST / HTTP/1.1
Host: micktestpost.s3.eu-north-1.amazonaws.com
Content-Type: multipart/form-data; boundary=abc
Content-Length: 10000

--abc
Content-Disposition: form-data; name="key";

THE_KEY
--abc
Content-Disposition: form-data; name="whatever";

unfinished body
EOF
```

Then I wait until AWS times out and I receive

```bash
HTTP/1.1 400 Bad Request
Transfer-Encoding: chunked
Connection: close

168
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>RequestTimeout</Code><Message>Your socket connection to the server was not read from or written to within the timeout period. Idle connections will be closed.</Message>...</Error>
0
```
It took 40s: `0,00s user 0,00s system 0% cpu 40,153 total`

Now if you put that field under a field with name `file` it only lasts for 20s.

And finish the body properly with the final closing boundary `--abc--` then it finish the request and doesn't care about the big Content-Length

Reply:
I wasn't able to replicate; I get an immediate empty response
```
➜  playground cat <<"EOF" | time nc -C test-bucket-post-acl.s3.us-east-1.amazonaws.com 80
POST / HTTP/1.1
Host: test-bucket-post-acl.s3.us-east-1.amazonaws.com
Content-Type: multipart/form-data; boundary=abc
Content-Length: 10000

--abc
Content-Disposition: form-data; name="key";

THE_KEY
--abc
Content-Disposition: form-data; name="whatever";

unfinished body
EOF
nc -C test-bucket-post-acl.s3.us-east-1.amazonaws.com 80  0.00s user 0.01s system 4% cpu 0.285 total
```

My bucket has the following perms to allow all traffic
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "Stmt1684751811099",
            "Effect": "Allow",
            "Principal": {
                "AWS": "*"
            },
            "Action": [
                "s3:GetObject",
                "s3:PutObject"
            ],
            "Resource": "arn:aws:s3:::test-bucket-post-acl/*"
        }
    ]
}
```

What might I be missing?

Reply:
Depending on your system you need to check the options for `nc`, on mac maybe you need to use `-c` instead of `-C` for CRLF, you need to make sure you don't shutdown the connection right after EOF and keep it open.

Reply:
see https://github.com/scality/cloudserver/pull/5601#discussion_r1690995359 I'm closing this